drivers/i2c/i2c_gd32.c

```diff
@@ -0,0 +1,684 @@
+	}
+}
+
+static void i2c_gd32_handle_btc(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+
+	if (data->current->flags & I2C_MSG_READ) {
+		uint32_t counter = 0U;
+
+		switch (data->xfer_len) {
+		case 2:
+			/*
+			 * Stop condition must be generated before reading the
+			 * last two bytes.
+			 */
+			if ((data->current->flags & I2C_MSG_STOP) &&
+			    (data->is_stopped == 0U)) {
+				data->is_stopped = 1U;
+				I2C_CTL0(cfg->reg) |= I2C_CTL0_STOP;
+			}
+
+			for (counter = 2U; counter > 0; counter--) {
+				data->xfer_len--;
+				i2c_gd32_xfer_read(data, cfg);
+			}
+
+			k_sem_give(&data->sync_sem);
+
+			break;
+		case 3:
+			I2C_CTL0(cfg->reg) &= ~I2C_CTL0_ACKEN;
+
+			data->xfer_len--;
+			i2c_gd32_xfer_read(data, cfg);
+
+			break;
+		default:
+			i2c_gd32_handle_rbne(dev);
+			break;
+		}
+	} else {
+		i2c_gd32_handle_tbe(dev);
+	}
+}
+
+static void i2c_gd32_handle_addsend(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+
+	if (data->is_restart == 1U) {
+		I2C_STAT1(cfg->reg);
+
+		data->is_restart = 0U;
+		data->current->flags &= ~I2C_MSG_RW_MASK;
+		data->current->flags |= I2C_MSG_READ;
+		I2C_CTL0(cfg->reg) |= I2C_CTL0_START;
+
+		return;
+	}
+
+	if ((data->current->flags & I2C_MSG_READ) &&
+	    (data->xfer_len <= 2U)) {
+		I2C_CTL0(cfg->reg) &= ~I2C_CTL0_ACKEN;
+	}
+
+	I2C_STAT1(cfg->reg);
+
+	if ((data->xfer_len == 1U) &&
+	    (data->current->flags & I2C_MSG_STOP) &&
+	    (data->current->flags & I2C_MSG_READ)) {
+		data->is_stopped = 1U;
+		I2C_CTL0(cfg->reg) |= I2C_CTL0_STOP;
+	}
+
+}
+
+static void i2c_gd32_event_isr(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+	uint32_t stat = I2C_STAT0(cfg->reg);
+
+	if (stat & I2C_STAT0_SBSEND) {
+		if (data->current->flags & I2C_MSG_READ) {
+			I2C_DATA(cfg->reg) = (data->addr1 << 1U) | 1U;
+		} else {
+			I2C_DATA(cfg->reg) = (data->addr1 << 1U) | 0U;
+		}
+	} else if (stat & I2C_STAT0_ADD10SEND) {
+		I2C_DATA(cfg->reg) = data->addr2;
+	} else if (stat & I2C_STAT0_ADDSEND) {
+		i2c_gd32_handle_addsend(dev);
+	} else if (stat & I2C_STAT0_BTC) {
+		i2c_gd32_handle_btc(dev);
+	} else if (stat & I2C_STAT0_RBNE) {
+		i2c_gd32_handle_rbne(dev);
+	} else if (stat & I2C_STAT0_TBE) {
+		i2c_gd32_handle_tbe(dev);
+	}
+}
+
+static void i2c_gd32_error_isr(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+	uint32_t stat = I2C_STAT0(cfg->reg);
+
+	if (stat & I2C_STAT0_BERR) {
+		I2C_STAT0(cfg->reg) &= ~I2C_STAT0_BERR;
+		data->is_berr = 1U;
+	}
+
+	if (stat & I2C_STAT0_LOSTARB) {
+		I2C_STAT0(cfg->reg) &= ~I2C_STAT0_LOSTARB;
+		data->is_lostarb = 1U;
+	}
+
+	if (stat & I2C_STAT0_AERR) {
+		I2C_STAT0(cfg->reg) &= ~I2C_STAT0_AERR;
+		data->is_aerr = 1U;
+	}
+
+	if (data->is_berr ||
```

Comment:
This is a place where perhaps using bit flags for the errors might make it easier to read, (if data->err != 0) { ... }

Reply:
Use errs to hold bus_err, lost_arb and ack_err.
done.